autonat.go

```diff
@@ -75,198 +70,227 @@ func NewAutoNAT(ctx context.Context, h host.Host, getAddrs GetAddrs) AutoNAT {
 		getAddrs = h.Addrs
 	}
 
-	emitUnknown, _ := h.EventBus().Emitter(new(event.EvtLocalRoutabilityUnknown))
-	emitPublic, _ := h.EventBus().Emitter(new(event.EvtLocalRoutabilityPublic))
-	emitPrivate, _ := h.EventBus().Emitter(new(event.EvtLocalRoutabilityPrivate))
+	subAddrUpdated, _ := h.EventBus().Subscribe(new(event.EvtLocalAddressesUpdated))
+
+	emitReachabilityChanged, _ := h.EventBus().Emitter(new(event.EvtLocalReachabilityChanged), eventbus.Stateful)
 
 	as := &AmbientAutoNAT{
-		ctx:      ctx,
-		host:     h,
-		getAddrs: getAddrs,
-		peers:    make(map[peer.ID][]ma.Multiaddr),
-		status:   NATStatusUnknown,
-
-		emitUnknown: emitUnknown,
-		emitPublic:  emitPublic,
-		emitPrivate: emitPrivate,
+		ctx:            ctx,
+		host:           h,
+		getAddrs:       getAddrs,
+		candidatePeers: make(chan network.Conn, 5),
+		observations:   make(chan autoNATResult, 1),
+
+		subAddrUpdated: subAddrUpdated,
+
+		emitReachabilityChanged: emitReachabilityChanged,
 	}
+	as.status.Store(autoNATResult{network.ReachabilityUnknown, nil})
 
 	h.Network().Notify(as)
 	go as.background()
 
 	return as
 }
 
-func (as *AmbientAutoNAT) Status() NATStatus {
-	as.mx.Lock()
-	defer as.mx.Unlock()
-	return as.status
+// Status returns the AutoNAT observed reachability status.
+func (as *AmbientAutoNAT) Status() network.Reachability {
+	s := as.status.Load().(autoNATResult)
+	return s.Reachability
 }
 
-func (as *AmbientAutoNAT) updateStatus(s NATStatus) {
-	as.status = s
-	switch s {
-	case NATStatusUnknown:
-		as.emitUnknown.Emit(event.EvtLocalRoutabilityUnknown{})
-	case NATStatusPublic:
-		as.emitPublic.Emit(event.EvtLocalRoutabilityPublic{})
-	case NATStatusPrivate:
-		as.emitPrivate.Emit(event.EvtLocalRoutabilityPrivate{})
-	}
+func (as *AmbientAutoNAT) emitStatus() {
+	status := as.status.Load().(autoNATResult)
+	as.emitReachabilityChanged.Emit(event.EvtLocalReachabilityChanged{Reachability: status.Reachability})
 }
 
+// PublicAddr returns the publicly connectable Multiaddr of this node if one is known.
 func (as *AmbientAutoNAT) PublicAddr() (ma.Multiaddr, error) {
-	as.mx.Lock()
-	defer as.mx.Unlock()
-
-	if as.status != NATStatusPublic {
+	s := as.status.Load().(autoNATResult)
+	if s.Reachability != network.ReachabilityPublic {
 		return nil, errors.New("NAT Status is not public")
 	}
 
-	return as.addr, nil
+	return s.address, nil
 }
 
 func (as *AmbientAutoNAT) background() {
 	// wait a bit for the node to come online and establish some connections
 	// before starting autodetection
-	select {
-	case <-time.After(AutoNATBootDelay):
-	case <-as.ctx.Done():
-		return
-	}
+	delay := AutoNATBootDelay
+
+	addrUpdatedChan := as.subAddrUpdated.Out()
+	defer as.subAddrUpdated.Close()
+	defer as.emitReachabilityChanged.Close()
 
 	for {
-		as.autodetect()
+		select {
+		// new connection occured.
+		case conn := <-as.candidatePeers:
```

Comment:
Let's fix the name of this channel. We're no longer using it for candidate peers.